File: dotnet/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example1/AirTrafficControlTower.cs
namespace MediatorDesignPattern.Example1;

/// <summary>
/// Concrete Mediator - Air Traffic Control Tower
/// Coordinates all communication between aircraft
/// </summary>
public class AirTrafficControlTower : IMediator
{
    private readonly List<Aircraft> _aircrafts = new();
    private readonly Queue<Aircraft> _landingQueue = new();
    private readonly Queue<Aircraft> _takeoffQueue = new();
    private bool _runwayAvailable = true;

    public void RegisterAircraft(Aircraft aircraft)
    {
        _aircrafts.Add(aircraft);
        Console.WriteLine($"[TOWER] {aircraft.CallSign} ({aircraft.AircraftType}) registered to the system.");
    }

    public void SendMessage(string message, Aircraft sender)
    {
        // Relay message to all aircraft except the sender
        foreach (var aircraft in _aircrafts.Where(a => a != sender))
        {
            aircraft.Receive($"{sender.CallSign}: {message}");
        }
    }

    public void RequestLanding(Aircraft aircraft)
    {
        if (_runwayAvailable && _landingQueue.Count == 0)
        {
            _runwayAvailable = false;
            Console.WriteLine($"[TOWER] {aircraft.CallSign}, cleared for landing. Runway is clear.");
            
            // Notify other aircraft
            foreach (var other in _aircrafts.Where(a => a != aircraft))
            {
                other.Receive($"[TOWER] {aircraft.CallSign} is landing, please hold.");
            }

            // Simulation: Runway will be available after 3 seconds
            Task.Delay(3000).ContinueWith(_ =>
            {
                _runwayAvailable = true;
                Console.WriteLine($"[TOWER] {aircraft.CallSign} landing completed. Runway is clear.");
                ProcessNextInQueue();
            });
        }
        else
        {
            _landingQueue.Enqueue(aircraft);
            Console.WriteLine($"[TOWER] {aircraft.CallSign}, runway is busy. You are #{_landingQueue.Count} in queue.");
            aircraft.Receive($"[TOWER] Please hold, you will be notified when it's your turn.");
        }
    }

    public void RequestTakeoff(Aircraft aircraft)
    {
        if (_runwayAvailable && _takeoffQueue.Count == 0 && _landingQueue.Count == 0)
        {
            _runwayAvailable = false;
            Console.WriteLine($"[TOWER] {aircraft.CallSign}, cleared for takeoff. Have a safe flight!");
            
            // Notify other aircraft
            foreach (var other in _aircrafts.Where(a => a != aircraft))
            {
                other.Receive($"[TOWER] {aircraft.CallSign} is taking off.");
            }

            // Simulation: Runway will be available after 2 seconds
            Task.Delay(2000).ContinueWith(_ =>
            {
                _runwayAvailable = true;
                Console.WriteLine($"[TOWER] {aircraft.CallSign} takeoff completed. Runway is clear.");
                ProcessNextInQueue();
            });
        }
        else
        {
            _takeoffQueue.Enqueue(aircraft);
            Console.WriteLine($"[TOWER] {aircraft.CallSign}, runway is busy or landing has priority. You are #{_takeoffQueue.Count} in queue.");
            aircraft.Receive($"[TOWER] Please hold, you will be notified when it's your turn.");
        }
    }

    private void ProcessNextInQueue()
    {
        // Landing has priority
        if (_landingQueue.Count > 0)
        {
            var nextAircraft = _landingQueue.Dequeue();
            Console.WriteLine($"[TOWER] Calling next aircraft: {nextAircraft.CallSign}");
            RequestLanding(nextAircraft);
        }
        else if (_takeoffQueue.Count > 0)
        {
            var nextAircraft = _takeoffQueue.Dequeue();
            Console.WriteLine($"[TOWER] Calling next aircraft: {nextAircraft.CallSign}");
            RequestTakeoff(nextAircraft);
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example1/src/AirTrafficControlTower.ts
import { IMediator } from './IMediator';
import { Aircraft } from './Aircraft';

const sleep = (ms: number) => new Promise<void>(resolve => setTimeout(resolve, ms));

export class AirTrafficControlTower implements IMediator {
  private readonly _aircrafts: Aircraft[] = [];
  private readonly _landingQueue: Aircraft[] = [];
  private readonly _takeoffQueue: Aircraft[] = [];
  private _runwayAvailable: boolean = true;

  registerAircraft(aircraft: Aircraft): void {
    this._aircrafts.push(aircraft);
    console.log(`[TOWER] ${aircraft.callSign} (${aircraft.aircraftType}) registered to the system.`);
  }

  sendMessage(message: string, sender: Aircraft): void {
    for (const aircraft of this._aircrafts.filter(a => a !== sender)) {
      aircraft.receive(`${sender.callSign}: ${message}`);
    }
  }

  requestLanding(aircraft: Aircraft): void {
    if (this._runwayAvailable && this._landingQueue.length === 0) {
      this._clearForLanding(aircraft);
    } else {
      this._landingQueue.push(aircraft);
      console.log(`[TOWER] ${aircraft.callSign}, runway is busy. You are #${this._landingQueue.length} in queue.`);
      aircraft.receive(`[TOWER] Please hold, you will be notified when it's your turn.`);
    }
  }

  requestTakeoff(aircraft: Aircraft): void {
    if (this._runwayAvailable && this._takeoffQueue.length === 0 && this._landingQueue.length === 0) {
      this._clearForTakeoff(aircraft);
    } else {
      this._takeoffQueue.push(aircraft);
      console.log(`[TOWER] ${aircraft.callSign}, runway is busy or landing has priority. You are #${this._takeoffQueue.length} in queue.`);
      aircraft.receive(`[TOWER] Please hold, you will be notified when it's your turn.`);
    }
  }

  private _clearForLanding(aircraft: Aircraft): void {
    this._runwayAvailable = false;
    console.log(`[TOWER] ${aircraft.callSign}, cleared for landing. Runway is clear.`);
    for (const other of this._aircrafts.filter(a => a !== aircraft)) {
      other.receive(`[TOWER] ${aircraft.callSign} is landing, please hold.`);
    }
    setTimeout(() => {
      this._runwayAvailable = true;
      console.log(`[TOWER] ${aircraft.callSign} landing completed. Runway is clear.`);
      this._processNextInQueue();
    }, 3000);
  }

  private _clearForTakeoff(aircraft: Aircraft): void {
    this._runwayAvailable = false;
    console.log(`[TOWER] ${aircraft.callSign}, cleared for takeoff. Have a safe flight!`);
    for (const other of this._aircrafts.filter(a => a !== aircraft)) {
      other.receive(`[TOWER] ${aircraft.callSign} is taking off.`);
    }
    setTimeout(() => {
      this._runwayAvailable = true;
      console.log(`[TOWER] ${aircraft.callSign} takeoff completed. Runway is clear.`);
      this._processNextInQueue();
    }, 2000);
  }

  private _processNextInQueue(): void {
    if (this._landingQueue.length > 0) {
      const next = this._landingQueue.shift()!;
      console.log(`[TOWER] Calling next aircraft: ${next.callSign}`);
      this._clearForLanding(next);
    } else if (this._takeoffQueue.length > 0) {
      const next = this._takeoffQueue.shift()!;
      console.log(`[TOWER] Calling next aircraft: ${next.callSign}`);
      this._clearForTakeoff(next);
    }
  }
}
